### PR TITLE
vsphere: Reconcile region and zone labels on Machine

### PR DESCRIPTION
**Pull request by bison**
This reconciles the labels on Machine objects used for region and zone
information.  In order to do so, the vSphere cloud provider must be
configured with the names of tags identifying this information.
    
  https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/zones.html

### CODE REVIEW

**Comment by enxebre** [comment]
can we move the vendor commit into its own?

**Comment by bison** [comment]
> can we move the vendor commit into its own?

Done.

File: pkg/controller/vsphere/util.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+
+	"gopkg.in/gcfg.v1"
+
```
Comment (enxebre) [inline pkg/controller/vsphere/util.go]: extra spaces here

Comment (bison) [reply pkg/controller/vsphere/util.go]: I intentionally separated it from the k8s related imports.

File: pkg/controller/vsphere/reconciler.go [hunk 2]
```diff
@@ -184,6 +188,10 @@ func (r *Reconciler) reconcileMachineWithCloudState(vm *virtualMachine) error {
 	klog.V(3).Infof("%v: reconciling machine with cloud state", r.machine.GetName())
 	// TODO: reconcile task
 
+	if err := r.reconcileRegionAndZoneLabels(vm); err != nil {
+		klog.Errorf("Failed to reconcile region and zone labels: %v", err)
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler.go]: don't we want to error here?
we might want to adapt `reconcileMachineWithCloudState` to return NewAggregate(errlist []error)

Comment (bison) [reply pkg/controller/vsphere/reconciler.go]: We can, but this seems totally optional. I thought I would rather have a working machine without region and zone labels (that are only used by the column printers in the CRD) than to fail here, but I honestly don't have a strong opinion.

File: pkg/controller/vsphere/util.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	if err := client.Get(context.Background(), infraName, infra); err != nil {
+		return nil, err
+	}
+
+	return infra, nil
+}
+
+func getVSphereConfig(client runtimeclient.Client) (*vSphereConfig, error) {
+	infra, err := getInfrastructure(client)
+	if err != nil {
```
Comment (enxebre) [inline pkg/controller/vsphere/util.go]: do you think we could skip this and assume the given name for the configmap -> https://github.com/openshift/installer/blob/a76a6b34092d99835d0e947ae9343d231f38f7e1/pkg/asset/manifests/cloudproviderconfig.go#L80

Comment (bison) [reply pkg/controller/vsphere/util.go]: We could, but I think this is not too complicated, and it would be less correct to hard code the name.

**Comment by bison** [comment]
/hold cancel

This seems to be working now:
```
$ kubectl -n vsphere get machines
NAME               PHASE         TYPE   REGION      ZONE         AGE
bison-vsphere-1a   Provisioned          us-east-1   us-east-1a   2m12s
bison-vsphere-1b   Provisioned          us-east-1   us-east-1b   104s
```

File: cmd/vsphere/main.go [hunk 4]
```diff
@@ -50,9 +51,14 @@ func main() {
 	// Initialize machine actuator.
 	machineActuator := machine.NewActuator(machine.ActuatorParams{
 		Client:        mgr.GetClient(),
+		RestConfig:    mgr.GetConfig(),
```
Comment (enxebre) [inline cmd/vsphere/main.go]: I think we could pass `mgr.GetAPIReader()` here and rely on it further down to fetch objects beyond the cache client namespace

Comment (enxebre) [reply cmd/vsphere/main.go]: https://kubernetes.slack.com/archives/CAR30FCJZ/p1580118306075000

File: install/0000_30_machine-api-operator_09_rbac.yaml [hunk 5]
```diff
@@ -11,6 +11,22 @@ metadata:
   name: machine-api-controllers
   namespace: openshift-machine-api
 
+---
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (enxebre) [inline install/0000_30_machine-api-operator_09_rbac.yaml]: this is superseding https://github.com/openshift/machine-api-operator/pull/473/files#diff-2043d8125934ffff48ca0cb4d630543bR376 can we drop it here?

File: pkg/controller/vsphere/util.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+type vSphereConfig struct {
+	Labels struct {
+		Zone   string `gcfg:"zone"`
+		Region string `gcfg:"region"`
+	}
+}
+
+func getInfrastructure(cfg *rest.Config) (*configv1.Infrastructure, error) {
+	if cfg == nil {
+		return nil, errors.New("No REST config")
```
Comment (enxebre) [inline pkg/controller/vsphere/util.go]: nit: use lowercase for passing through error

File: pkg/controller/vsphere/util.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	infrav1 := clientset.ConfigV1().Infrastructures()
+
+	return infrav1.Get(globalInfrastuctureName, metav1.GetOptions{})
+}
+
+func getVSphereConfig(cfg *rest.Config) (*vSphereConfig, error) {
+	if cfg == nil {
+		return nil, errors.New("No REST config")
```
Comment (enxebre) [inline pkg/controller/vsphere/util.go]: nit: use lowercase for passing through error

File: pkg/controller/vsphere/session/session.go [hunk 8]
```diff
@@ -156,3 +165,20 @@ func (s *Session) GetTask(ctx context.Context, taskRef string) (*mo.Task, error)
 	}
 	return &obj, nil
 }
+
+func (s *Session) WithRestClient(ctx context.Context, f func(c *rest.Client) error) error {
```
Comment (enxebre) [inline pkg/controller/vsphere/session/session.go]: the passed parameter `func(c *rest.Client)` does not seem to be used?

Comment (enxebre) [reply pkg/controller/vsphere/session/session.go]: nvm, I see is required to ensure the `f(c)` call below.
Do we need the login/logout calls?
What's the reason why we can not simplify by just passing `rest.NewClient(s.Client.Client)` where needed with out this wrapper? e.g `vm.getRegionAndZone(rest.NewClient(s.Client.Client), regionLabel, zoneLabel)`

Comment (bison) [reply pkg/controller/vsphere/session/session.go]: The function returns the result of applying `f()`: `return f(c)`

Comment (bison) [reply pkg/controller/vsphere/session/session.go]: > Do we need the login/logout calls?

Yes, or else the REST client is not authenticated.

> What's the reason why we can not simplify by just passing rest.NewClient(s.Client.Client) where needed with out this wrapper? e.g vm.getRegionAndZone(rest.NewClient(s.Client.Client), regionLabel, zoneLabel)

We could pass a rest client directly to that method, but we'd have to authenticate it. Doing that in any place that might need it doesn't seem like a simplification.

File: cmd/vsphere/main.go (hunk 4 above)
Comment (bison) [reply cmd/vsphere/main.go]: Fixed.

File: install/0000_30_machine-api-operator_09_rbac.yaml (hunk 5 above)
Comment (bison) [reply install/0000_30_machine-api-operator_09_rbac.yaml]: Fixed.

File: pkg/controller/vsphere/util.go (hunk 6 above)
Comment (bison) [reply pkg/controller/vsphere/util.go]: Fixed. This bit is gone.

File: pkg/controller/vsphere/util.go (hunk 7 above)
Comment (bison) [reply pkg/controller/vsphere/util.go]: Fixed. This bit is gone.

File: pkg/controller/vsphere/actuator.go [hunk 9]
```diff
@@ -24,19 +25,23 @@ const (
...
+	apiReader     runtimeclient.Reader
 	eventRecorder record.EventRecorder
 }
 
 // ActuatorParams holds parameter information for Actuator.
 type ActuatorParams struct {
 	Client        runtimeclient.Client
+	APIReader     runtimeclient.Reader
 	EventRecorder record.EventRecorder
+	RestConfig    *rest.Config
```
Comment (enxebre) [inline pkg/controller/vsphere/actuator.go]: we can drop this one now right?

**Comment by enxebre** [comment]
other than https://github.com/openshift/machine-api-operator/pull/471/commits/131ebebbf86f729de57db18a6bf5b706082cda62#r371222161 lgtm
/approve

File: pkg/controller/vsphere/actuator.go (hunk 9 above)
Comment (bison) [reply pkg/controller/vsphere/actuator.go]: Oops. Yeah. Fixed.

File: pkg/controller/vsphere/reconciler.go [hunk 10]
```diff
@@ -465,6 +511,66 @@ type virtualMachine struct {
...
+			category, err := tagsMgr.GetCategory(vm.Context, tag.CategoryID)
+			if err != nil {
+				klog.Errorf("Failed to get tag category: %v", err)
+				return nil, err
+			}
+
+			switch {
+			case regionLabel != "" && category.Name == regionLabel:
+				result[regionKey] = tag.Name
+				klog.Infof("%s has region tag (%s) with value %s",
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler.go]: I don't know how many ancestors a VM might have, but could this log be quite noisy? Similar with the one below? 
Looks like this codepath is expected to execute every reconcile, for each ancestor of the VM that has the right label right? Should this maybe be a more verbose log level?

Comment (bison) [reply pkg/controller/vsphere/reconciler.go]: True. I've made it `V(2)` and now return early once we've found both tags, which is something it should have been doing anyway.

**Pull request merged**